### PR TITLE
[VCDA-3533] Delete the vApp if and only if rdeId (matches) present in the vApp

### PR DESCRIPTION
**Pull request by ymo24**
* During delete cluster operation, delete the vApp if and only if vApp has the ```InfraId``` as metadata property.
* Created a generic validation function in vApp.go to check whether a key-value pair is matched in vApp's metadata.
* Test Done:
1. If vAppmetadataUpdated is true, deletion process will be in stuck.
2. In current version, vAppmetadataUpdate will not be reset until the deletion process is done.
3. If InfraId is mismatched, deletion process will be in stuck and customers could see error hint in logs.

* To summarize, customers have choice to keep any component they want (vApp, VMs, load balance components).

---
This change is [Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/110)

### CODE REVIEW

**Review by Anirudh9794 — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/110)* status: all files reviewed, 1 unresolved discussion (waiting on @ymo24)

___
*[`pkg/vcdclient/vapp.go` line 174 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/110#-N1KiLwh5EJvHp0j4XAS:-N1KiLwh5EJvHp0j4XAT:b-snmjvo) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/d1b133420f0a926d8b4f7d3a1130c99ea7e81c24/pkg/vcdclient/vapp.go#L174)):*
> ```Go
> }
> 
> func (vdc *VdcManager) ValidateMetadata(vApp *govcd.VApp, key string, value string) (bool, error) {
> ```

I think this function is doing too much work. 

Can we change the function signature to
```
func (vdc *VdcManager) GetMetadataByKey(vApp *govcd.VApp, key string) (value string, err error)
```
and make the comparison for the value within the controllers?

**Review by ymo24 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/110)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @Anirudh9794)

___
*[`pkg/vcdclient/vapp.go` line 174 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/110#-N1KiLwh5EJvHp0j4XAS:-N1KxDbG3BrDppqAKuLD:bxr9fpp) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/d1b133420f0a926d8b4f7d3a1130c99ea7e81c24/pkg/vcdclient/vapp.go#L174)):*
<details><summary><i>Previously, Anirudh9794 wrote…</i></summary><blockquote>

I think this function is doing too much work. 

Can we change the function signature to
```
func (vdc *VdcManager) GetMetadataByKey(vApp *govcd.VApp, key string) (value string, err error)
```
and make the comparison for the value within the controllers?
</blockquote></details>

Already pushed the commits. Thank you!

**Comment by Anirudh9794** [comment]
*[`controllers/vcdmachine_controller.go` line 846 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/110#-N1irj5r7FNk7mx_wpMM:-N1irj5r7FNk7mx_wpMN:b5pazqj) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/3f23bcd56bfdb3283221678208ebd3d4f1ad2576/controllers/vcdmachine_controller.go#L846)):*
> ```Go
> 			return ctrl.Result{}, errors.Errorf("Error occurred during the machine deletion; Metadata not found in vApp")
> 		}
> 		metadataInfraId, err := vdcManager.GetMetadataByKey(vApp, CapvcdInfraId)
> ```

Check for
```
if err != nil {}
```
is missing here

**Review by Anirudh9794 — APPROVED** [review]
:lgtm: minor comment

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/110#-:-N1irquu1IF5Sq8cgbnV:basp5ut)* status: all files reviewed, 2 unresolved discussions (waiting on @Anirudh9794, @sahithi, and @ymo24)

File: controllers/vcdmachine_controller.go [hunk 1]
```diff
@@ -839,6 +839,20 @@ func (r *VCDMachineReconciler) reconcileDelete(ctx context.Context, cluster *clu
 		}
 	}
 	if vApp != nil {
+		// Delete the vApp if and only if rdeId (matches) present in the vApp
```
Comment (sahithi) [inline controllers/vcdmachine_controller.go:842]: Nit: delete the VM...

**Review by sahithi — APPROVED** [review]
:lgtm: minor comment

> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/110#-:-N1kusPu-Y6ioBuqYN8n:basp5ut)* status: all files reviewed, 3 unresolved discussions (waiting on @Anirudh9794, @lzichong, and @ymo24)

**Review by ymo24 — COMMENTED** [review]
Thank you very much.

> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/110#-:-N1nz1PR7EoEyTMk4pZU:b-6utjl9)* status: 1 of 3 files reviewed, 2 unresolved discussions (waiting on @Anirudh9794 and @lzichong)

___
*[`controllers/vcdmachine_controller.go` line 846 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/110#-N1irj5r7FNk7mx_wpMM:-N1ix361CVtvRvQoM2kE:bnputky) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/3f23bcd56bfdb3283221678208ebd3d4f1ad2576/controllers/vcdmachine_controller.go#L846)):*
<details><summary><i>Previously, Anirudh9794 wrote…</i></summary><blockquote>

Check for
```
if err != nil {}
```
is missing here
</blockquote></details>

Done. Thanks
